packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts

```diff
@@ -1,20 +1,358 @@
+            'Rust router real routing performance',
+        );
+    }
+
+    const routesAndSamples = _.zip(allSourcesRustRoute, samplesAndNativeOrdersWithResults);
+
+    const adjustedFills: Fill[] = [];
+    const totalRoutedAmount = BigNumber.sum(...allSourcesRustRoute);
+
+    const scale = input.dividedBy(totalRoutedAmount);
+    for (const [routeInput, routeSamplesAndNativeOrders] of routesAndSamples) {
+        if (!routeInput || !routeSamplesAndNativeOrders) {
+            continue;
+        }
+        // TODO(kimpers): [TKR-241] amounts are sometimes clipped in the router due to precisions loss for number/f64
+        // we can work around it by scaling it and rounding up. However now we end up with a total amount of a couple base units too much
+        const rustInput = BigNumber.min(
+            new BigNumber(routeInput).multipliedBy(scale).integerValue(BigNumber.ROUND_CEIL),
+            input,
+        );
+
+        const current = routeSamplesAndNativeOrders[routeSamplesAndNativeOrders.length - 1];
+        if (!isDexSample(current)) {
+            const nativeFill = nativeOrdersToFills(
+                side,
+                [current],
+                rustInput,
+                opts.outputAmountPerEth,
+                opts.inputAmountPerEth,
+                fees,
+            )[0];
+            // NOTE: For Limit/RFQ orders we are done here. No need to scale output
+            adjustedFills.push(nativeFill);
+            continue;
+        }
+
+        // NOTE: For DexSamples only
+        let fill = createFill(current);
+        const routeSamples = routeSamplesAndNativeOrders as Array<DexSample<FillData>>;
+        // Descend to approach a closer fill for fillData which may not be consistent
+        // throughout the path (UniswapV3) and for a closer guesstimate at
+        // gas used
+        for (let k = routeSamples.length - 1; k >= 0; k--) {
+            if (k === 0) {
+                fill = createFill(routeSamples[0]);
+            }
+            if (rustInput.isGreaterThan(routeSamples[k].input)) {
+                // Between here and the previous fill
+                // HACK: Use the midpoint between the two
+                const left = routeSamples[k];
+                const right = routeSamples[k + 1];
+                if (left && right) {
+                    const leftPrice = left.output.dividedBy(left.input);
+                    const rightPrice = right.output.dividedBy(right.input);
+                    const scaledPrice = leftPrice
+                        .minus(rightPrice)
+                        .dividedBy(left.input.minus(right.input))
+                        .times(rustInput.minus(right.input))
+                        .plus(rightPrice);
+                    const output = scaledPrice
+                        .times(rustInput)
+                        .decimalPlaces(0, side === MarketOperation.Sell ? BigNumber.ROUND_FLOOR : BigNumber.ROUND_CEIL);
```

Comment:
I am not entirely sure what we're doing here. Can we get a comment on the formula?

Reply:
Is this all equivalent to...?
```ts
                    const stride = right.input.minus(left.input);
                    const rateBetween = right.output.minus(left.output).dividedBy(stride);
                    const output = rustInput
                        .minus(left.input)
                        .dividedBy(stride)
                        .times(rateBetween)
                        .plus(left.input)
                        .decimalPlaces(...);
```

Reply:
I could be wrong but I don't think that is exactly equivalent 😅 

I don't know why we use price here to interpolate (shamelessly stole it from Jacobs PoC that used the midpoint price), I have cleaned this up and refactored to do interpolate on the input and output amounts instead with some useful comments. Lmkwyt!

Reply:
I just want to be sure we're interpolating using the price *between* the two samples, and not the difference between each sample's price (from 0).